src/RelayedTransactionValidator.ts

```diff
@@ -0,0 +1,95 @@
+import log from 'loglevel';
+import type { Transaction } from 'ethers';
+import type { ContractInteractor } from '@rsksmart/rif-relay-common';
+import type {
+  EnvelopingConfig,
+  RelayTransactionRequest,
+  DeployTransactionRequest,
+} from '@rsksmart/rif-relay-common/dist/src';
+
+export class RelayedTransactionValidator {
+  private readonly contractInteractor: ContractInteractor;
+
+  private readonly config: EnvelopingConfig;
+
+  constructor(
+    contractInteractor: ContractInteractor,
+    config: EnvelopingConfig
+  ) {
+    this.contractInteractor = contractInteractor;
+    this.config = config;
+  }
+
+  /**
+   * Decode the signed transaction returned from the Relay Server, compare it to the
+   * requested transaction and validate its signature.
+   */
+  validateRelayResponse(
+    request: RelayTransactionRequest | DeployTransactionRequest,
+    { v, r, s, to, data, value, from, nonce }: Transaction,
+    relayWorker: string
+  ): void {
+    log.info('validateRelayResponse - Transaction is', {
+      v,
+      r,
+      s,
+      to,
+      from,
+      data,
+      value,
+      nonce,
+    });
+
+    if (!to) {
+      throw Error('Transaction has no recipient address');
+    }
+
+    if (!from) {
+      throw Error('Transaction has no signer');
+    }
+
+    const isDeploy = (request as DeployTransactionRequest).relayRequest.request
+      .recoverer
+      ? true
+      : false;
+
+    const relayRequestAbiEncode = isDeploy
+      ? this.contractInteractor.encodeDeployCallABI(
+          (request as DeployTransactionRequest).relayRequest,
+          request.metadata.signature
+        )
+      : this.contractInteractor.encodeRelayCallABI(
+          (request as RelayTransactionRequest).relayRequest,
+          request.metadata.signature
+        );
+
+    // TODO: the relayServer encoder returns zero-length buffer for nonce=0.`
```

Comment:
I think we missed the part related to this comment. In the previous version, there was a specific conversion from zero-length buffer to number.
```
const receivedNonce =
                transaction.nonce.length === 0
                    ? 0
                    : transaction.nonce.readUIntBE(
                          0,
                          transaction.nonce.byteLength
                      );
```

Reply:
We may add a test to mock such response from the server.

Reply:
btw @antomor the code could be cleaner as:
```
const { nonce } = transaction;
const receivedNonce = nonce.length && nonce.readUIntBE(0, nonce.byteLength);
```

Reply:
![image](https://user-images.githubusercontent.com/12575626/201270172-5a1152f3-dc18-4ac1-add0-a7b5e7116d8d.png)
I might need some help with these. Currently, nonce is a number, which means it has no property length, byteLength or much less readUintBE. Maybe there's a more appropiate type than just Transaction?

Reply:
this doesn't apply anymore since we are using Ethers Transactions